File: packages/sdk-py/af_stack/jobs.py
"""``suite.jobs.*`` — background job operations.

Maps to the REST surface defined in ``apps/dashboard/src/lib/api.ts`` under
the Phase 5 Jobs section:

==================================  ====================================
SDK call                            Endpoint
==================================  ====================================
:func:`enqueue`                     ``POST   /api/v1/jobs``
:func:`get`                         ``GET    /api/v1/jobs/{id}``
:func:`retry`                       ``POST   /api/v1/jobs/{id}/retry``
:func:`list`                        ``GET    /api/v1/jobs``
==================================  ====================================

The :class:`Job` and :class:`JobList` Pydantic models mirror the zod schemas
in ``lib/api.ts`` exactly — keep them in sync when the contract changes.

Admin operations (job *definitions*, cron management, etc.) live in the
post-v1 admin SDK. They're intentionally absent here.
"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from . import _http

JobState = Literal[
    "available",
    "running",
    "completed",
    "discarded",
    "cancelled",
    "retryable",
    "scheduled",
    "pending",
]


class JobError(BaseModel):
    """One row from a job's ``errors`` array."""

    model_config = ConfigDict(extra="allow")

    at: str
    attempt: int
    error: str


class Job(BaseModel):
    """A background job row.

    Mirrors ``JobSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    id: str
    name: str
    args: Any = None
    state: JobState
    tenant_id: str | None = None
    attempt: int
    max_attempts: int
    scheduled_at: str
    enqueued_at: str
    attempted_at: str | None = None
    finalized_at: str | None = None
    errors: list[JobError] | None = None


class JobList(BaseModel):
    """Paginated list of jobs.

    Mirrors ``JobListSchema`` in ``apps/dashboard/src/lib/api.ts``.
    """

    model_config = ConfigDict(extra="allow")

    jobs: list[Job] = Field(default_factory=list)
    total: int = 0
    has_more: bool = False


async def enqueue(
    name: str,
    args: dict[str, Any] | None = None,
    *,
    scheduled_at: datetime | None = None,
    max_attempts: int | None = None,
) -> Job:
    """Enqueue a background job and return the persisted row.

    ``name`` is the job definition name (e.g. ``"send-welcome-email"``).
    ``args`` is the JSON-serialisable argument payload forwarded verbatim
    to the worker. ``scheduled_at`` defers execution until the given time;
    when omitted the runtime queues immediately.
    """
    if not name:
        raise ValueError("job name must be a non-empty string")
    payload: dict[str, Any] = {
        "name": name,
        "args": args if args is not None else {},
    }
    if scheduled_at is not None:
        payload["scheduled_at"] = _serialize_dt(scheduled_at)
    if max_attempts is not None:
        if max_attempts < 1 or max_attempts > 50:
            raise ValueError("max_attempts must be between 1 and 50")
        payload["max_attempts"] = max_attempts
    body = await _http.request_json("POST", "/jobs", json=payload)
    return Job.model_validate(body or {})


async def get(id: str) -> Job:
    """Fetch a job by id."""
    if not id:
        raise ValueError("job id must be a non-empty string")
    body = await _http.request_json("GET", f"/jobs/{id}")
    return Job.model_validate(body or {})


async def retry(id: str) -> Job:
    """Force-retry a job (resets state to ``available``)."""
    if not id:
        raise ValueError("job id must be a non-empty string")
    body = await _http.request_json("POST", f"/jobs/{id}/retry")
    return Job.model_validate(body or {})


async def list(  # noqa: A001 — mirrors the JS `jobs.list()` ergonomics
    *,
    name: str | None = None,
    state: JobState | str | None = None,
    tenant: str | None = None,
    limit: int = 50,
    offset: int = 0,
) -> JobList:
    """List jobs with optional filters.

    Filter parameter names match the dashboard API (``name``, ``state``,
    ``tenant``).
    """
    params: dict[str, Any] = {"limit": limit, "offset": offset}
    if name is not None:
        params["name"] = name
    if state is not None:
        params["state"] = state
    if tenant is not None:
        params["tenant"] = tenant
    body = await _http.request_json("GET", "/jobs", params=params)
    return JobList.model_validate(body or {})


def _serialize_dt(dt: datetime) -> str:
    """ISO-8601 with a ``Z`` suffix for naive UTC datetimes."""
    iso = dt.isoformat()
    # If the datetime is naive, treat as UTC for predictability.
    if dt.tzinfo is None:
        return iso + "Z"
    return iso


__all__ = [
    "Job",
    "JobError",
    "JobList",
    "JobState",
    "enqueue",
    "get",
    "list",
    "retry",
]

File: packages/sdk-ts/src/jobs.ts
// SPDX-License-Identifier: Apache-2.0

// suite.jobs.* — background job operations.
//
// Endpoint paths and JSON shapes are the canonical contract from
// `apps/dashboard/src/lib/api.ts` (the Phase 5 Jobs section). Schemas
// below mirror the zod schemas there; if the dashboard contract moves,
// move these together.
//
// Public API stays camelCase; we translate snake_case <-> camelCase at the
// wire boundary using helpers in `_http.ts`.

import { z } from "zod"
import { request, type HttpOptions } from "./_http.js"

// ---------- shared schemas (mirror lib/api.ts) ----------

export const JobStateSchema = z.enum([
  "available",
  "running",
  "completed",
  "discarded",
  "cancelled",
  "retryable",
  "scheduled",
  "pending",
])
export type JobState = z.infer<typeof JobStateSchema>

const JobErrorSchema = z.object({
  at: z.string(),
  attempt: z.number(),
  error: z.string(),
})
export type JobError = z.infer<typeof JobErrorSchema>

export const JobSchema = z.object({
  id: z.string(),
  name: z.string(),
  args: z.unknown(),
  state: JobStateSchema,
  tenantId: z.string().nullable(),
  attempt: z.number(),
  maxAttempts: z.number(),
  scheduledAt: z.string(),
  enqueuedAt: z.string(),
  attemptedAt: z.string().nullable(),
  finalizedAt: z.string().nullable(),
  errors: z.array(JobErrorSchema).nullable(),
})
export type Job = z.infer<typeof JobSchema>

export const JobListSchema = z.object({
  jobs: z.array(JobSchema),
  total: z.number(),
  hasMore: z.boolean(),
})
export type JobList = z.infer<typeof JobListSchema>

// ---------- public API ----------

export interface EnqueueJobOptions extends HttpOptions {
  /** ISO 8601 / Date — defer execution until this moment. */
  scheduledAt?: Date | string
  /** Override the worker's default attempt cap (1–50). */
  maxAttempts?: number
}

export interface ListJobsOptions extends HttpOptions {
  name?: string
  state?: JobState | string
  tenant?: string
  limit?: number
  offset?: number
}

/** Enqueue a background job. Returns the persisted row. */
export async function enqueue(
  name: string,
  args: unknown,
  opts: EnqueueJobOptions = {},
): Promise<Job> {
  if (typeof name !== "string" || name.length === 0) {
    throw new Error("job name must be a non-empty string")
  }
  const { scheduledAt, maxAttempts, ...http } = opts
  if (maxAttempts !== undefined && (maxAttempts < 1 || maxAttempts > 50)) {
    throw new Error("maxAttempts must be between 1 and 50")
  }
  const body: Record<string, unknown> = {
    name,
    args: args ?? {},
  }
  if (scheduledAt !== undefined) {
    body.scheduled_at = serializeDate(scheduledAt)
  }
  if (maxAttempts !== undefined) {
    body.max_attempts = maxAttempts
  }
  const raw = await request<unknown>("POST", "/jobs", body, http)
  return JobSchema.parse(camelizeJob(raw))
}

/** Fetch a job by id. */
export async function get(id: string, opts: HttpOptions = {}): Promise<Job> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("job id must be a non-empty string")
  }
  const raw = await request<unknown>(
    "GET",
    `/jobs/${encodeURIComponent(id)}`,
    null,
    opts,
  )
  return JobSchema.parse(camelizeJob(raw))
}

/** Force-retry a job (resets state to `available`). */
export async function retry(id: string, opts: HttpOptions = {}): Promise<Job> {
  if (typeof id !== "string" || id.length === 0) {
    throw new Error("job id must be a non-empty string")
  }
  const raw = await request<unknown>(
    "POST",
    `/jobs/${encodeURIComponent(id)}/retry`,
    null,
    opts,
  )
  return JobSchema.parse(camelizeJob(raw))
}

/** List jobs with optional filters. */
export async function list(opts: ListJobsOptions = {}): Promise<JobList> {
  const { name, state, tenant, limit = 50, offset = 0, ...http } = opts
  const qs = new URLSearchParams()
  if (name !== undefined) qs.set("name", name)
  if (state !== undefined) qs.set("state", state)
  if (tenant !== undefined) qs.set("tenant", tenant)
  qs.set("limit", String(limit))
  qs.set("offset", String(offset))
  const raw = await request<unknown>("GET", `/jobs?${qs.toString()}`, null, http)
  return JobListSchema.parse(camelizeJobList(raw))
}

// ---------- helpers ----------

function serializeDate(value: Date | string): string {
  if (value instanceof Date) return value.toISOString()
  return value
}

/**
 * Translate the snake_case wire envelope into the camelCase shape declared
 * by `JobSchema`, *without* touching `args` (which is agent-defined and
 * forwarded verbatim). Mirrors the OPAQUE_FIELDS pattern in agents.ts.
 */
function camelizeJob(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  const out: Record<string, unknown> = {}
  for (const [k, v] of Object.entries(src)) {
    const ck = camelKey(k)
    if (ck === "args") {
      out.args = v
    } else {
      out[ck] = v
    }
  }
  return out
}

function camelizeJobList(raw: unknown): unknown {
  if (raw === null || typeof raw !== "object" || Array.isArray(raw)) return raw
  const src = raw as Record<string, unknown>
  const jobs = Array.isArray(src.jobs) ? src.jobs.map(camelizeJob) : src.jobs
  return {
    jobs,
    total: src.total,
    hasMore: src.has_more ?? src.hasMore,
  }
}

function camelKey(key: string): string {
  return key.replace(/_([a-z0-9])/g, (_, c: string) => c.toUpperCase())
}

/** Namespace object — the shape `suite.jobs` is built from. */
export const jobs = {
  enqueue,
  get,
  retry,
  list,
} as const